pkg/operator/ceph/object/shared_pools.go

```diff
@@ -0,0 +1,532 @@
+package object
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path"
+	"reflect"
+	"sort"
+
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+)
+
+const (
+	defaultPlacementName           = "default"
+	defaultPlacementCephConfigName = "default-placement"
+	defaultPlacementStorageClass   = "STANDARD"
+)
+
+func IsNeedToCreateObjectStorePools(sharedPools cephv1.ObjectSharedPoolsSpec) bool {
+	for _, pp := range sharedPools.PoolPlacements {
+		if pp.Name == defaultPlacementName {
+			// No need to create pools. External pools from default placement will be used
+			return false
+		}
+	}
+	if sharedPools.MetadataPoolName != "" && sharedPools.DataPoolName != "" {
+		// No need to create pools. Shared pools will be used
+		return false
+	}
+	return true
+}
+
+func validatePoolPlacements(placements []cephv1.PoolPlacementSpec) error {
+	names := make(map[string]struct{}, len(placements))
+	for _, p := range placements {
+		if _, ok := names[p.Name]; ok {
+			return fmt.Errorf("invalidObjStorePoolCofig: invalid placement %s: placement names must be unique", p.Name)
+		}
+		names[p.Name] = struct{}{}
+		if err := validatePoolPlacementStorageClasses(p.StorageClasses); err != nil {
+			return fmt.Errorf("invalidObjStorePoolCofig: invalid placement %s: %w", p.Name, err)
+		}
+	}
+	return nil
+}
+
+func validatePoolPlacementStorageClasses(scList []cephv1.PlacementStorageClassSpec) error {
+	names := make(map[string]struct{}, len(scList))
+	for _, sc := range scList {
+		if sc.Name == defaultPlacementStorageClass {
+			return fmt.Errorf("invalid placement StorageClass %s: %s name is reserved", sc.Name, defaultPlacementStorageClass)
+		}
+		if _, ok := names[sc.Name]; ok {
+			return fmt.Errorf("invalid placement StorageClass %s: name must be unique", sc.Name)
+		}
+		names[sc.Name] = struct{}{}
+	}
+	return nil
+}
+
+func adjustZonePlacementPools(zone ZoneJSON, spec cephv1.ObjectSharedPoolsSpec) (bool, error) {
+	name, err := zone.getString("name")
+	if err != nil {
+		return false, fmt.Errorf("unable to get zone name: %w", err)
+	}
+
+	old, err := zone.getPoolPlacements()
+	if err != nil {
+		return false, fmt.Errorf("unable to get pool placements for zone %s: %w", name, err)
+	}
+	oldMap := make(map[string]ZonePlacementPool, len(old))
+	for _, pp := range old {
+		oldMap[pp.Key] = pp
+	}
+	// map spec to new zone pools config
+	updated := toZonePlacementPools(spec, name)
+	newMap := make(map[string]ZonePlacementPool, len(updated))
+	for i, pp := range updated {
+		if oldPl, ok := oldMap[pp.Key]; ok {
+			// preserve original values which are not covered by pool placement CRD
+			pp.Val.IndexType = oldPl.Val.IndexType
+			pp.Val.InlineData = oldPl.Val.InlineData
+			updated[i].Val.IndexType = oldPl.Val.IndexType
+			updated[i].Val.InlineData = oldPl.Val.InlineData
```

Comment:
i'm guessing that "index type" corresponds to [enum BucketIndexType](https://github.com/ceph/ceph/blob/64f5bff3/src/rgw/rgw_bucket_layout.h#L30-L32) where 0=Normal and 1=Indexless

inline data is documented in a Note under https://docs.ceph.com/en/latest/radosgw/placement/#adding-a-placement-target

Reply:
> Is it possible to manage the zone and leave the fields blank (unspecified)? Would that be a good or bad idea?

i take this to imply that you're building up a json document and creating the zone with `radosgw-admin zone set`? i think this is a bad idea in general. you probably do need `zone set` to customize the pool names, but you should probably run `zone create` first so it can fill in all the defaults, then use `zone get` -> change pool names -> `zone set`

my main concern with `zone set` by itself is that new ceph releases will add new fields for pool names and enable new [zone features](https://docs.ceph.com/en/latest/radosgw/zone-features/) by default that rook won't keep up-to-date

Reply:
Understood, and that is exactly what we are considering here. As always, your input is very clear and helpful.

It sounds like this (below) will be the safe approach for Rook internally. I think is the path Artem has been on already, so I just want to make sure we are all clearly targeting the same thing.

1. Check if zone exists; if not, `zone create` (this ensures good defaults)
2. Get zone config JSON, and copy it fully to Rook's internal structure
3. In this case, we are configuring pool placements, so Rook then overwrites the pool placement config(s) with what Rook generates as the necessary config.

For part 3, it will be best to overwrite only the fields that we know Rook needs to set and that Rook can confidently take full ownership of.

@cbodley If some part of this still sounds off, let us know. 

@arttor, I'd like to keep discussing this. You've shown that you are an insightful developer who writes quality code, and I'm grateful.

My interpretation of the code we're reviewing here is that the implication is that Rook is taking full ownership of the pool placement config including these index type and inline data fields. I think the adjustment we might need to consider is to have Rook copy all fields without caveat (which should include index-type/inline-data implicitly) to make sure Rook doesn't clobber any fields added in the future.

If my interpretation seems wrong, let me know.

Reply:
> 1. Check if zone exists; if not, zone create (this ensures good defaults)

might be simpler to issue the `zone create` unconditionally and ignore EEXIST errors

> @cbodley If some part of this still sounds off, let us know.

sounds good

Reply:
just one note regarding creating a zone. Rook objecstore controller always calls [createNonMultisiteStore()](https://github.com/rook/rook/blob/master/pkg/operator/ceph/object/objectstore.go#L433-L469) before configuring zone. This method makes sure that zone, zonegroup, and realm are always presented.